crates/rolldown/src/stages/link_stage/tree_shaking/include_statements.rs

```diff
@@ -48,34 +50,80 @@ impl LinkStage<'_> {
       runtime_id: self.runtime.id(),
       // used_exports_info_vec: &mut used_exports_info_vec,
       metas: &self.metas,
-      used_symbol_refs: &mut self.used_symbol_refs,
+      used_symbol_refs: &mut used_symbol_refs,
       options: self.options,
     };
 
-    self.entries.iter().for_each(|entry| {
+    let (user_defined_entries, mut dynamic_entries): (Vec<_>, Vec<_>) =
+      std::mem::take(&mut self.entries)
+        .into_iter()
+        .partition(|item| matches!(item.kind, EntryPointKind::UserDefined));
+
+    user_defined_entries
+      .iter()
+      .filter(|entry| matches!(entry.kind, EntryPointKind::UserDefined))
+      .for_each(|entry| {
+        let module = match &self.module_table.modules[entry.id] {
+          Module::Normal(module) => module,
+          Module::External(_module) => {
+            // Case: import('external').
+            return;
+          }
+        };
+        let meta = &self.metas[entry.id];
+        meta.referenced_symbols_by_entry_point_chunk.iter().for_each(|symbol_ref| {
+          include_symbol(context, *symbol_ref);
+        });
+        include_module(context, module);
+      });
+
+    if self.options.is_hmr_enabled() {
+      // HMR runtime contains statements with side effects, they are referenced by other modules via global variables.
+      // So we need to manually include them here.
+      if let Some(runtime_module) = self.module_table.modules[self.runtime.id()].as_normal() {
+        runtime_module.stmt_infos.iter_enumerated().for_each(|(stmt_info_id, stmt_info)| {
+          if stmt_info.side_effect {
+            include_statement(context, runtime_module, stmt_info_id);
+          }
+        });
+      }
+    }
+
+    let mut unused_record_idxs = vec![];
+
+    dynamic_entries.sort_by_key(|item| self.module_table.modules[item.id].exec_order());
+
+    dynamic_entries.retain(|entry| {
```

Comment:
We can't compute the unused dynamic entry AOT, because some dynamic entries maybe descendants of a dynamic entry module.